privacyidea/api/lib/decorators.py

```diff
@@ -0,0 +1,119 @@
+
+    """
+    def __init__(self, function, request, position="pre", options=()):
+        """
+        :param function: This is the function the is to be called
+        :type function: function
+        :param request: The original request object, that needs to be passed
+        :type request: Request Object
+        :param options: Options which are passed to the decorator function
+        :type options: dict
+        :param position: The behavior of the decorator. With "post", it executes
+        after the request. With "pre" (default), the API function itself is wrapped
+        by the decorator.
+        """
+        self.function = function
+        self.request = request
+        self.position = position
+        self.options = options
+
+    def __call__(self, wrapped_function):
+        """
+        This decorates the given function according to the given position.
+
+        :param wrapped_function: The function, that is decorated.
+        :type wrapped_function: API function
+        :return: None
+        """
+        @functools.wraps(wrapped_function)
+        def function_wrapper(*args, **kwds):
+            if self.position == "post":
+                response = wrapped_function(*args, **kwds)
+                return self.function(self.request,
+                                     response=response,
+                                     options=self.options,
+                                     *args, **kwds)
+            else:
+                self.function(self.request,
+                              options=self.options)
+                return wrapped_function(*args, **kwds)
+
+        return function_wrapper
+
+
+def g_add_serial(request, response=None, options=()):
+    """
+    This decorator function fetches the token serial and adds it to the flask
+    g object.
+    With a given response, this decorator checks for the serial in the response.
+    If no response is given, the decorator checks first in the request and in the
+    request path for the serial. By default the serial is expected at the third
+    position of the request path. Set the position of the serial in the URL by
+    specifying
+
+    options = {"serial_position": N}
+
+    N defaults to 3.
+    """
+    if response is None:
+        if "serial" not in g or not g.serial:
+            # try to get serial from request (also done by before_after.py)
+            g.serial = getParam(request.all_data, "serial")
+            if not g.serial:
+                # try to get serial from request path
+                if request.path:
+                    if "serial_position" not in options:
+                        options = {"serial_position": 3}
+                    path_elements = [x for x in request.path.split('/') if x]
+                    if len(path_elements) > options["serial_position"] + 1:
+                        g.serial = path_elements[options["serial_position"] + 1]
+        return True
```

Comment:
I would like this method to get the position directly in the ``option`` parameter.

Then we could do it like this:

~~~~python
if request.path:
    position = option or 2 # 2 is the default?
    path_elements = [x for x in request.path.split('/') if x]
    if len(path_elements) >= position:
        g.serial = path_elements[position -1]